esp_custom_part_ota/src/esp_custom_part_ota.c

```diff
@@ -0,0 +1,281 @@
+        if (!ctx->backup_partition) {
+            ESP_LOGW(TAG, "No backup partition found");
+        }
+    }
+    ctx->need_erase = 1;
+    ctx->backup_len = 0;
+
+    esp_custom_part_ota_handle_t handle = (esp_custom_part_ota_handle_t)ctx;
+    return handle;
+}
+
+esp_err_t esp_custom_part_ota_write(esp_custom_part_ota_handle_t handle, const void *data, size_t size)
+{
+    if (!handle) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *ctx = (esp_custom_part_ota_t *)handle;
+    if (!ctx->update_partition) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_err_t ret = ESP_FAIL;
+    if (ctx->need_erase) {
+        ret = esp_partition_erase_range(ctx->update_partition, 0, ctx->update_partition->size);
+        if (ret == ESP_OK) {
+            ctx->need_erase = 0;
+            ESP_LOGI(TAG, "Successfully erased update partition");
+        } else {
+            return ret;
+        }
+    }
+    ret = esp_partition_write(ctx->update_partition, ctx->wrote_size, data, size);
+    if (ret == ESP_OK) {
+        ctx->wrote_size += size;
+    }
+    return ret;
+}
+
+esp_err_t esp_custom_part_ota_end(esp_custom_part_ota_handle_t handle)
+{
+    if (!handle) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *ctx = (esp_custom_part_ota_t *)handle;
+    if (ctx->wrote_size == 0) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    free(ctx);
+    return ESP_OK;
+}
+
+esp_err_t esp_custom_part_ota_abort(esp_custom_part_ota_handle_t handle)
+{
+    if (!handle) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *ctx = (esp_custom_part_ota_t *)handle;
+    free(ctx);
+    return ESP_OK;
+}
+
+esp_err_t esp_custom_part_ota_set_backup_partition(esp_custom_part_ota_handle_t handle, const esp_partition_t *backup_partition)
+{
+    if (!handle || !backup_partition) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *ctx = (esp_custom_part_ota_t *)handle;
+    ctx->backup_partition = backup_partition;
+    return ESP_OK;
+}
+
+esp_err_t esp_custom_part_ota_partition_backup(esp_custom_part_ota_handle_t handle, size_t backup_size)
+{
+    if (!handle) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *ctx = (esp_custom_part_ota_t *)handle;
+    if (!ctx->backup_partition) {
+        ESP_LOGE(TAG, "Backup partition not set. Cannot backup");
+        return ESP_FAIL;
+    }
+    ctx->backup_len = (backup_size == 0) ? ctx->update_partition->size : backup_size;
+    if (ctx->backup_len > ctx->backup_partition->size) {
+        ESP_LOGE(TAG, "Backup partition size smaller than data to be backed up");
+        return ESP_FAIL;
+    }
+    size_t block_size = 4096;
+    uint8_t *data = calloc(block_size, sizeof(uint8_t));
+    if (!data) {
+        ESP_LOGE(TAG, "Could not allocate memory for data");
+        return ESP_ERR_NO_MEM;
+    }
+
+    esp_err_t ret = esp_partition_erase_range(ctx->backup_partition, 0, ctx->backup_len);
```

Comment:
Can we allow this API to be executed twice for 2 different partitions? If not, how can we detect if there exist a backup for any previous partition? Do we need some API to invalidate the previous backup here?

Reply:
Oh right. I didn't think of this case. I think we can store some information in NVS regarding this and iterate over the data to check if some other backup is stored in the partition.